osf/models/preprint_service.py

```diff
@@ -260,3 +323,729 @@ def _send_preprint_confirmation(self, auth):
+            return False
+
+        return self.is_published or (auth.user and self.has_permission(auth.user, 'read'))
+
+    def can_edit(self, auth=None, user=None):
+        """Return if a user is authorized to edit this node.
+        Must specify one of (`auth`, `user`).
+
+        :param Auth auth: Auth object to check
+        :param User user: User object to check
+        :returns: Whether user has permission to edit this node.
+        """
+        if not auth and not user:
+            raise ValueError('Must pass either `auth` or `user`')
+        if auth and user:
+            raise ValueError('Cannot pass both `auth` and `user`')
+        user = user or auth.user
+        if auth:
+            is_api_node = auth.api_node == self
+        else:
+            is_api_node = False
+        return (
+            (user and self.has_permission(user, 'write')) or is_api_node
+        )
+
+    # TODO: Remove save parameter
+    def add_permission(self, user, permission, save=False):
+        """Grant permission to a user.
+
+        :param User user: User to grant permission to
+        :param str permission: Permission to grant
+        :param bool save: Save changes
+        :raises: ValueError if user already has permission
+        """
+
+        if not self.has_permission(user, permission):
+            self.get_group(permission).user_set.add(user)
+        elif self.has_permission(user, permission):
+            raise ValueError('User already has permission {0}'.format(permission))
+        if save:
+            self.save()
+
+    # TODO: Remove save parameter
+    def remove_permission(self, user, permission, save=False):
+        """Revoke permission from a user.
+
+        :param User user: User to revoke permission from
+        :param str permission: Permission to revoke
+        :param bool save: Save changes
+        :raises: ValueError if user does not have permission
+        """
+        if user in self.get_group(permission).user_set:
+            self.get_group(permission).user_set.remove(user)
+        else:
+            raise ValueError('User does not have permission {0}'.format(permission))
+        if save:
+            self.save()
+
+    def clear_permissions(self, user):
+        for name in self.group_names:
+            if name == 'admin':
+                if self.get_group('admin').user_set.exclude(id=user.id).exists():
+                    self.remove_permission(user, 'admin')
+                else:
+                    raise PreprintStateError('Must have at least one registered admin contributor')
+            self.remove_permission(user, name)
+
+    def get_visible(self, user):
+        try:
+            contributor = self.preprintcontributor_set.get(user=user)
+        except PreprintContributor.DoesNotExist:
+            raise ValueError(u'User {0} not in contributors'.format(user))
+        return contributor.visible
+
+    @property
+    def visible_contributors(self):
```

Comment:
Is this ever used? Any unused methods/properties should be removed.

Reply:
In the case for Node, it is used https://github.com/CenterForOpenScience/osf.io/blob/develop/api/base/pagination.py#L213. So this will be useful for PreprintContribtuorPagination.